pkgs/os-specific/darwin/install-binary-package/default.nix

```diff
@@ -0,0 +1,44 @@
+{ lib, stdenvNoCC, makeSetupHook, _7zz, libarchive }:
+
+{ appName ? "."
+, nativeBuildInputs ? []
+, sourceRoot ? "."
+, ...
+}@args:
+
+let
+  unpackDmgPkg = makeSetupHook {
+    name = "unpack-dmg-pkg";
+    propagatedBuildInputs = [ _7zz libarchive ];
+  } ./unpack-dmg-pkg.sh;
+in
+stdenvNoCC.mkDerivation (args // {
+  dontMakeSourcesWritable = args.dontMakeSourcesWritable or true;
+  dontPatch = args.dontPatch or true;
+  dontConfigure = args.dontConfigure or true;
+  dontBuild = args.dontBuild or true;
+  dontStrip = args.dontStrip or true;
+  dontFixup = args.dontFixup or true;
+  dontCheck = args.dontCheck or true;
+  doInstallCheck = args.doInstallCheck or false;
+
+  sourceRoot = "${args.pname}/${sourceRoot}";
```

Comment:
I think this might be a bit more involved than it seems and possibly a more elegant way exists to solve what the current approach solves.

Most of the binary sources unpack multiple files, a directory, or a one or more archive files (often named `Payload`). Yet some packages unpack a single binary, which made¹ the `unpackPhase` fail for me. Given the variety of binary source packages (dmg with file(s), pkg with file(s), dmg with archive(s), pkgs with archive(s)) it seems easier to unpack each "layer" and take it from there instead of trying to analyse what the "best" unpack strategy for a particular package is. Not that some packages also need multiple levels of unpacking, e.g. [`karabiner-elements`](https://github.com/NixOS/nixpkgs/blob/master/pkgs/os-specific/darwin/karabiner-elements/default.nix#L16-L24).

So my idea was to simple put all unpacked files into a directory named `args.pname` to satisfy the directory-"requirement"—in which a directory must be present afterwards—of the `unpackPhase`.
So introducing `subSourceRoot` might not provide the solution that is needed…

Have I been able to explain my thinking around this well?
What are alternative approaches for this challenge?

────────────
¹ Maybe this has been addressed as I cannot not reproduce the issue; possibly my quick tests for this comment have been too simple to not trigger the error. I'll keep my 👀  peeled.

Reply:
> So my idea was to simple put all unpacked files into a directory named `args.pname` to satisfy the directory-"requirement"—in which a directory must be present afterwards—of the `unpackPhase`.
>  So introducing subSourceRoot might not provide the solution that is needed…

That makes sense. `subSourceRoot` doesn't seem to scratch the itch.

Then, how about setting `sourceRoot = "${args.pname}"` by default, and tell people to always set `sourceRoot` to `"${pname}"` or its sub-directory inside Nixpkgs manual or comment block?

The current implementation will result in developer setting `sourceRoot = "subdir"` and see `sourceRoot is hello-bin/subdir` in the build log. Setting `sourceRoot = "${pname}/subdir"` and see `sourceRoot is hello-bin/subdir` avoids such inconsistency.

From the search result inside the Nixpkgs code base with regular expression `sourceRoot =.*sourceRoot`, no build helpers provided in Nixpkgs change `sourceRoot` transparently like this. The only exception is [`buildIntegration`](https://github.com/NixOS/nixpkgs/blob/master/pkgs/tools/networking/dd-agent/integrations-core.nix#L54) defined locally inside a `let ... in` block.

I could only express my point of view instead of requesting a change, as I'm nither a committer nor a Darwin user. The design choice will be made by you and the committers.